src/Dict.h

```diff
@@ -138,6 +147,71 @@ class DictEntry {
 
 } // namespace detail
 
+class DictIterator {
+public:
+	using value_type = detail::DictEntry;
+	using reference = detail::DictEntry&;
+	using pointer = detail::DictEntry*;
+	using difference_type = std::ptrdiff_t;
+	using iterator_category = std::bidirectional_iterator_tag;
+
+	DictIterator() = default;
+	DictIterator(const Dictionary* d, detail::DictEntry* begin, detail::DictEntry* end);
+	~DictIterator();
+
+	reference operator*() { return *curr; }
+	pointer operator->() { return curr; }
+
+	DictIterator& operator++();
+	DictIterator operator++(int) { auto temp(*this); ++*this; return temp; }
+
+	bool operator==( const DictIterator& that ) const { return curr == that.curr; }
+	bool operator!=( const DictIterator& that ) const { return !(*this == that); }
+
+private:
+
+	Dictionary* dict = nullptr;
+	detail::DictEntry* curr = nullptr;
+	detail::DictEntry* end = nullptr;
+};
+
+class RobustDictIterator : public std::enable_shared_from_this<RobustDictIterator> {
+public:
+	using value_type = detail::DictEntry;
+	using reference = detail::DictEntry&;
+	using pointer = detail::DictEntry*;
+	using difference_type = std::ptrdiff_t;
+	using iterator_category = std::bidirectional_iterator_tag;
+
+	RobustDictIterator() : curr(nullptr) {}
+	RobustDictIterator(Dictionary* d);
+ 	~RobustDictIterator();
+
+	void Complete();
+	bool Active() { return inserted != nullptr; }
+	std::shared_ptr<RobustDictIterator> GetPtr() { return shared_from_this(); }
```

Comment:
There's awkwardness/downsides to `shared_ptr` / `shared_from_this` dependence:

* Implies only valid way to use these objects is if they're created within a `shared_ptr`, except nothing currently enforces that.

* If they're only supposed to be used as `shared_ptr`, not sure what help `GetPtr()` is since "you've already got a shared_ptr, just make a copy".

* Current API is at odds with implications since `begin_robust()` returns a `shared_ptr` and `end_robust()` does not.  There's not currently any of our code where an end-iterator ever ends up doing an invalid `shared_from_this()`, but there is a cost in terms of just requiring users to understand the complexity of it -- whether a given method is ok to use depends on how the object was originally created and/or whether it's been wrapped in `shared_ptr`.

* I didn't verify, but using `shared_ptr` here may introduce a new way to leak memory rather than make things safer: upon creation of a robust-iterator, the `Dictionary` stores a reference to it as a `shared_ptr` and that reference is not removed until the dictionary is completely-iterated or else an explicit `Complete()` call is made on the iterator.  E.g. breaking a loop early and letting your iterator-pointer go out of scope without doing a `Complete()` on it manually is a leak.

What are the concerns with having Dictionary store a list of raw-pointers-to-iterators ?  Seems simpler (probably do also need `RobustDictIterator` copy/move ctor/assignments in that case, but can't think of much more to it).

Reply:
I would have preferred to just return a normal object there instead of a pointer, except that TableVal stores an iterator for a while in order to stretch expiration over a number of passes. Storing an object there instead of storing a pointer bloats memory quite a bit. The reason why end_robust doesn't return a pointer is to avoid having to dereference it every loop to check for equality.

I'll dig back into this and see if I can find a better way, but that at least gives some context.

Reply:
> If they're only supposed to be used as shared_ptr, not sure what help GetPtr() is since "you've already got a shared_ptr, just make a copy".

`GetPtr()` isn't actually being used so I can remove it. It was a leftover from some other experiments I was working on and I never took it out.

> Current API is at odds with implications since begin_robust() returns a shared_ptr and end_robust() does not.

Like I said, I'd prefer to not have to return a pointer at all, but due to storing the iterator for a long period of time it saves a considerable amount of memory. I can see making `end_robust` return the same type though, just for consistency.

> I didn't verify, but using shared_ptr here may introduce a new way to leak memory rather than make things safer: upon creation of a robust-iterator, the Dictionary stores a reference to it as a shared_ptr and that reference is not removed until the dictionary is completely-iterated or else an explicit Complete() call is made on the iterator. E.g. breaking a loop early and letting your iterator-pointer go out of scope without doing a Complete() on it manually is a leak.

I'm fairly sure that the existing robust iteration would do the same thing. I copied the new `dict robust iteration` test back into the current Dict.cc, and added an if statement to drop out of one of the loops early. ASan says that it leaks the IterCookie. This is because without reaching the end of NextEntry and the end of the iteration, then the cookie never gets deleted and it never gets removed from the vector in the dictionary. You have to know to explicitly call `Dictionary::StopIteration()` in that case for it to not leak. I'm not saying that it's not a bug here too, but it's one that already exists in the code.

The primary reason why I opted for a shared_ptr vs a raw pointer was an ownership question. If both `TableVal` and `Dictionary` store the same raw pointer, which is the true "owner" of that pointer? Which is truly responsible for disposing of it? I'd be inclined to say `TableVal` since it's the one that created it and is directly using it, but Dictionary is holding onto it for good reasons: in order to insert/remove elements from it during iterations.

Reply:
> Like I said, I'd prefer to not have to return a pointer at all, but due to storing the iterator for a long period of time it saves a considerable amount of memory.

If `begin_robust()` returned a regular iterator object, what would stop anyone from heap-allocating a copy-or-move-constructed iterator from that and storing the resulting pointer instead of the full object ?

> If both `TableVal` and `Dictionary` store the same raw pointer, which is the true "owner" of that pointer?

If `TableVal` wants to store a pointer to heap-allocated iterator for efficiency reasons, then it owns the memory (possibly `unique_ptr` even).

A `Dictionary` doesn't have an ownership relationship with iterators, just a referential one and the lifetime of that reference is automatically managed by the iterator destructor.  Also would say it's left up to user to ensure iterators don't outlive their associated container (not that shared iterator pointers help there anyway) -- think that's the same deal as it would be with standard-library stuff.

If it's mostly the memory-management-optics of storing pointers-to-iterators in `Dictionary` that are concerning, then maybe a `reference_wrapper` or something helps with that.

Reply:
I reworked begin_robust in the latest commit to return a regular object now if you want to take a look. There's some deprecation warnings that I need to resolve from restoring the methods in file_analysis, but this should be getting close now.